src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1576,26 +1563,58 @@ CitusHasBeenLoaded(void)
 			 */
 			DistPartitionRelationId();
 
-
 			/*
 			 * This needs to be initialized so we can receive foreign relation graph
 			 * invalidation messages in InvalidateForeignRelationGraphCacheCallback().
 			 * See the comments of InvalidateForeignKeyGraph for more context.
 			 */
 			DistColocationRelationId();
-
-			/*
-			 * We also reset citusVersionKnownCompatible, so it will be re-read in
-			 * case of extension update.
-			 */
-			citusVersionKnownCompatible = false;
 		}
+
+		MetadataCache.extensionLoaded = extensionLoaded;
 	}
 
 	return MetadataCache.extensionLoaded;
 }
 
 
+/*
+ * CitusHasBeenLoadedInternal returns true if the citus extension has been created
+ * in the current database and the extension script has been executed. Otherwise,
+ * it returns false.
+ */
+static bool
+CitusHasBeenLoadedInternal(void)
+{
+	Oid citusExtensionOid = InvalidOid;
+
+	if (IsBinaryUpgrade)
+	{
+		/* never use Citus logic during pg_upgrade */
+		return false;
+	}
+
+	citusExtensionOid = get_extension_oid("citus", true);
```

Comment:
Caching this seems to drop another 10% overhead. Without your PR, it takes 12s for me, with your PR 950ms, with caching this 860ms.